clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -734,11 +740,11 @@ private ClusterResourceListeners configureClusterResourceListeners(Serializer<K>
     private int partition(ProducerRecord<K, V> record, byte[] serializedKey, byte[] serializedValue, Cluster cluster) {
         Integer partition = record.partition();
         if (partition != null) {
-            List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int lastPartition = partitions.size() - 1;
+            int lastPartition = cluster.partitionsForTopic(record.topic()).size() - 1;
             // they have given us a partition, use it
             if (partition < 0 || partition > lastPartition) {
```

Comment:
I think we can dispense with the `partition < 0` case earlier. Maybe we can check it first thing in `doSend` before we even try to fetch metadata? For that case, `IllegalArgumentException` is probably reasonable because we will never have a negative partition.